File: src/ast/tree/statement/functionDeclaration.ts
import type { Pass } from "../../../pass/pass";
import type { Block } from "../../node/block";
import { ExecutionResult, Statement } from "../../node/statement";
import type { Function } from "../function/function";

export class FunctionDeclaration extends Statement {
    constructor(public func: Function, parent: Block) {
        super(parent);
    }

    public execute(): ExecutionResult {
        this.parent!.symbolTable.set(this.func.name, this.func);
        return ExecutionResult.ContinueEvaluation;
    }

    public visit(pass: Pass): void {
        pass.visitFunctionDeclaration(this);
    }
}

File: src/main/kotlin/me/honkling/miniscript/parser/ast/statement/FunctionDeclaration.kt
package me.honkling.miniscript.parser.ast.statement

import me.honkling.miniscript.diagnostic.Location
import me.honkling.miniscript.parser.ast.Block
import me.honkling.miniscript.parser.ast.function.Function
import me.honkling.miniscript.pass.Pass

class FunctionDeclaration(
    val function: Function,
    location: Location,
    parent: Block
) : Statement(location, parent) {
    override fun execute(): ExecutionResult {
        parent.miniScript.environment.lastFrame.symbolTable[function.name!!] = function
        return ExecutionResult.ContinueExecution
    }

    override fun accept(pass: Pass) {
        pass.visitFunctionDeclaration(this)
    }
}